internal/provider/resource_apply_module.go

```diff
@@ -0,0 +1,573 @@
+package provider
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/aws/smithy-go/ptr"
+	"github.com/google/uuid"
+	"github.com/hashicorp/terraform-plugin-framework/attr"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringplanmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-framework/types/basetypes"
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+	"github.com/martian-cloud/terraform-provider-tharsis/internal/modifiers"
+	tharsis "gitlab.com/infor-cloud/martian-cloud/tharsis/tharsis-sdk-go/pkg"
+	sdktypes "gitlab.com/infor-cloud/martian-cloud/tharsis/tharsis-sdk-go/pkg/types"
+)
+
+type doRunInput struct {
+	model     *ApplyModuleModel
+	doDestroy bool
+}
+
+type doRunOutput struct {
+	moduleVersion string
+}
+
+// appliedModuleInfo contains what information was available about the latest applied run.
+// One or both fields may be nil, in which case information was not available.
+type appliedModuleInfo struct {
+	moduleSource  *string
+	moduleVersion *string
+}
+
+const (
+	jobCompletionPollInterval = 5 * time.Second
+)
+
+var (
+	applyRunComment = "terraform-provider-tharsis" // must be var, not const, to take address
+)
+
+// RunVariableModel is used in apply modules to set Terraform and environment variables.
+type RunVariableModel struct {
+	Value    string `tfsdk:"value"`
+	Key      string `tfsdk:"key"`
+	Category string `tfsdk:"category"`
+	HCL      bool   `tfsdk:"hcl"`
+}
+
+// FromTerraform5Value converts a RunVariable from Terraform values to Go equivalent.
+// The digit in the name is necessary in order to pass the acceptance test.
+func (e *RunVariableModel) FromTerraform5Value(val tftypes.Value) error {
```

Comment:
I would recommend that we link a reference to the terraform [tftypes](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-go/tfprotov5/tftypes) docs that indicated we are implementing a required interface for the provider to use our custom type.